src/librustc_mir/transform/check_unsafety.rs

```diff
@@ -630,40 +670,79 @@ pub fn check_unsafety(tcx: TyCtxt<'_>, def_id: DefId) {
         tcx.unsafety_check_result(def_id.expect_local());
 
-    for &UnsafetyViolation { source_info, description, details, kind } in violations.iter() {
+    for &UnsafetyViolation { source_info, lint_root, description, details, kind } in
+        violations.iter()
+    {
         // Report an error.
+        let unsafe_fn_msg =
+            if unsafe_op_in_unsafe_fn_allowed(tcx, lint_root) { " function or" } else { "" };
+
         match kind {
             UnsafetyViolationKind::GeneralAndConstFn | UnsafetyViolationKind::General => {
+                // once
                 struct_span_err!(
                     tcx.sess,
                     source_info.span,
                     E0133,
-                    "{} is unsafe and requires unsafe function or block",
-                    description
+                    "{} is unsafe and requires unsafe{} block",
+                    description,
+                    unsafe_fn_msg,
                 )
                 .span_label(source_info.span, &*description.as_str())
                 .note(&details.as_str())
                 .emit();
             }
-            UnsafetyViolationKind::BorrowPacked(lint_hir_id) => {
+            UnsafetyViolationKind::BorrowPacked => {
                 if let Some(impl_def_id) = builtin_derive_def_id(tcx, def_id) {
                     tcx.ensure().unsafe_derive_on_repr_packed(impl_def_id);
                 } else {
                     tcx.struct_span_lint_hir(
                         SAFE_PACKED_BORROWS,
-                        lint_hir_id,
+                        lint_root,
                         source_info.span,
                         |lint| {
                             lint.build(&format!(
-                                "{} is unsafe and requires unsafe function or block (error E0133)",
-                                description
+                                "{} is unsafe and requires unsafe{} block (error E0133)",
+                                description, unsafe_fn_msg,
                             ))
                             .note(&details.as_str())
                             .emit()
                         },
                     )
                 }
             }
+            UnsafetyViolationKind::UnsafeFn => tcx.struct_span_lint_hir(
+                UNSAFE_OP_IN_UNSAFE_FN,
+                lint_root,
+                source_info.span,
+                |lint| {
+                    lint.build(&format!(
+                        "{} is unsafe and requires unsafe block (error E0133)",
+                        description,
+                    ))
+                    .span_label(source_info.span, &*description.as_str())
+                    .note(&details.as_str())
+                    .emit();
+                },
+            ),
+            UnsafetyViolationKind::UnsafeFnBorrowPacked => {
+                let lint = if tcx.lint_level_at_node(SAFE_PACKED_BORROWS, lint_root).0
```

Comment:
It took me a while to figure it out, but I think this is indeed the right test. I have two correlaries:

Ultimately, with `#[deny(unsafe_op_in_unsafe_fn)]`, I expect the behavior of a safe function and an unsafe function to be the same in terms of what warnings/errors I get. Therefore, if I have `#[warn(safe_packed_borrows)]`, I should get a warning in both contexts.

I also expect that `#[warn(unsafe_op_in_unsafe_fn)]` can't cause any new errors. Therefore, if you have `#[deny(safe_packed_borrws)]` and `#[warn(unsafe_op_in_unsafe_fn)]`, that should still only issue a warning.

Aaannnnd I guess a reasonable way to implement that is to do exactly what you have done. Very good.

Do we test all the combinations? I feel like we are not testing `deny(safe_packed_borrows)` combined with `warn(unsafe_op_in_unsafe_fn)`, but maybe I missed it. This bit of logic seems tricky enough that I think I'd like to see a file with each of the combinations in a separate function, though I guess that is 9 functions.

Also, can we add a comment here clarifying *why* this rule is setup the way it is? (My two correlaries above might be nice to paste in.)